eventauth/eventcontent.go

```diff
@@ -0,0 +1,280 @@
+package eventauth
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+)
+
+// creaateContent is the JSON content of a m.room.create event along with
+// the top level keys needed for auth.
+type createContent struct {
+	// We need the domain of the create event when checking federatebility.
+	senderDomain string
+	// We need the roomID to check that events are in the same room as the create event.
+	roomID string
+	// The "m.federate" flag tells us whether the room can be federated to other servers.
+	Federate *bool `json:"m.federate"`
+	// The creator of the room tells us what the default power levels are.
+	Creator string `json:"creator"`
+}
+
+// load the create event content from the create event in the auth events.
+func (c *createContent) load(authEvents AuthEvents) error {
+	createEvent, err := authEvents.Create()
+	if err != nil {
+		return err
+	}
+	if createEvent == nil {
+		return errorf("missing create event")
+	}
+	if err = json.Unmarshal(createEvent.Content, c); err != nil {
+		return errorf("unparsable create event content: %s", err.Error())
+	}
+	c.roomID = createEvent.RoomID
+	if c.senderDomain, err = domainFromID(createEvent.Sender); err != nil {
+		return err
+	}
+	return nil
+}
+
+// domainAllowed checks whether the domain is allowed in the room by the
+// "m.federate" flag.
+func (c *createContent) domainAllowed(domain string) error {
+	if domain == c.senderDomain {
+		// If the domain matches the domain of the create event
+		return nil
+	}
+	if c.Federate == nil || *c.Federate {
+		// The m.federate field defaults to true.
+		return nil
+	}
+	return errorf("room is unfederatable")
+}
+
+// idAllowed checks whether the domain of the ID is allowed in the room.
+func (c *createContent) idAllowed(id string) error {
+	domain, err := domainFromID(id)
+	if err != nil {
+		return err
+	}
+	return c.domainAllowed(domain)
+}
+
+// domainFromID returns everything after the first ":" character to extract
+// the domain part of a matrix ID.
+func domainFromID(id string) (string, error) {
+	// IDs have the format: SIGIL LOCALPART ":" DOMAIN
+	// Split on the first ":" character since the domain can contain ":"
+	// characters.
+	parts := strings.SplitN(id, ":", 2)
+	if len(parts) != 2 {
+		// The ID must have a ":" character.
+		return "", errorf("invalid ID: %q", id)
+	}
+	// Return everything after the first ":" character.
+	return parts[1], nil
+}
+
+// memberContent is the JSON content of a m.room.member event.
```

Comment:
Again, I'm struggling to conceptualise what `memberContent` is actually supposed to be representing. It sure isn't what you're stating here, because it's missing a bunch of fields from http://matrix.org/docs/spec/client_server/r0.2.0.html#m-room-member - given this is for *auth checks* I imagine you only care about the fields which are used for auth, so you should mention this in your description.